indexing-service/src/main/java/org/apache/druid/indexing/common/task/CompactionTask.java

```diff
@@ -550,16 +552,18 @@ private String createIndexTaskSpecId(int i)
         segmentProvider,
         lockGranularityInUse
     );
+
     final Map<DataSegment, File> segmentFileMap = pair.lhs;
     final List<TimelineObjectHolder<String, DataSegment>> timelineSegments = pair.rhs;
 
     if (timelineSegments.size() == 0) {
       return Collections.emptyList();
     }
 
-    // find metadata for interval
+    // find metadata for intervals with real data segments
     // queryableIndexAndSegments is sorted by the interval of the dataSegment
-    final List<NonnullPair<QueryableIndex, DataSegment>> queryableIndexAndSegments = loadSegments(
+    // Note that this list will contain null QueriableIndex values for tombstones
+    final List<Pair<QueryableIndex, DataSegment>> queryableIndexAndSegments = loadSegments(
```

Comment:
I haven't checked all of the call sites, so just asking, but given that from looking at this diff, it appears as those you adjusted all of the places that know about this list to start filtering out non-null values, what's the difference between just filtering them out here?

It would've been a lot nicer if there were just noop versions of these classes that could be used so that this logic doesn't have to change at all, but without doing that, filtering seems to be necessary.  But, why filter locally in all of the places instead of just once and reuse?

Reply:
Yes, fully agree. The code should be adapted eventually so that it returns no-op queriable indices or something along those lines, this would also help simplifying the query path. For now, this change is beneficial since now it will limit the filtering in the ingestion side to just the compaction task. Let me look into the compaction task to see if I can further simplify that code.

Reply:
The reason I put null values is to leave the current code flow as is as most as possible to mitigate the possibility of introducing a bug if refactoring. So now that list will also carry tombstones (thus nulls in the corresponding entry in the pair for a tombstone).